horovod/common/ops/adasum/adasum.h

```diff
@@ -0,0 +1,555 @@
+  // grad_buffer: holds the data to reduce and will hold the result.
+  // recv_buffer: must point to a buffer of the same size as grad_buffer.
+  // horovod_datatype: the element type of grad_buffer.
+  // tensor_counts: is a list of how many elements grad_buffer contains for each tensor
+  //                involved in the allreduce. It should contain a 0 if this rank holds
+  //                no data for the tensor (see start_level below for when this can happen).
+  // start_level: set to 1 to perform all levels of the operation. When set to n>1 the
+  //              first n-1 levels are skipped. This is useful when the communication inside
+  //              the node is implemented using another reduce-scatter algorithm, e.g. the
+  //              one in NCCL, which may be desireable on some hardware configurations. When
+  //              start_level>1, tensor_counts must be set according to the slices owned by
+  //              this rank.
+  // communicator: the communicator to reduce with.
+  // tag: a value used as the message tag for each send/recv in this algorithm. This is
+  //      useful for multithreaded scenarios. Remember to also create separate
+  //      reduction_comms instances when running with multiple threads.
+  // reduction_comms: pointer to an array of communicators for computing dot products and
+  //                  norms for Adasum. The communicators should include exactly the ranks
+  //                  that this rank has either directly or indirectly communicated with
+  //                  after each level of VHDD.
+  template <typename T>
+  void FusedAllreduce(std::vector<TensorTableEntry>& entries, 
+                      T* grad_buffer,
+                      T* recv_buffer,
+                      DataType horovod_datatype,
+                      std::vector<int>& tensor_counts,
+                      int start_level,
+                      Communicator_type communicator,
+                      int tag,
+                      Communicator_type* reduction_comms,
+                      HorovodGlobalState *global_state) {
+    int per_element_size = GetPerElementSize(horovod_datatype);
+    int rank = GetLocalRankWithComm(communicator);
+    int size = GetSizeWithComm(communicator);
+
+    if (IsPowerOfTwo(size) == false) {
+      throw std::logic_error(
+          "Running Adasum with non-power-of-two ranks is not supported yet.");
+    }
+    std::vector<std::vector<int>> nghrCountVec;
+    std::vector<double> normAndDots(tensor_counts.size()*3 * 2);
+
+    int nearest_power_2 = 1;
+    for (nearest_power_2 = 1; (nearest_power_2 << 1) <= size;
+        nearest_power_2 = (nearest_power_2 << 1)) {
+    }
+    int remaining_non_power_2 = size - nearest_power_2;
+    int level;
+
+    int nghrCountVec_index = 0;
+    int orgSize = size;
+    size = nearest_power_2;
+
+    int total_counts_sum = 0;
+    for (int i = 0; i < tensor_counts.size(); i++)
```

Comment:
Compiler warnings: please change all `int i = 0` in loops over vectors to `size_t = 0`.